design/secrets_provider_milestone1_2_design.md

```diff
@@ -0,0 +1,380 @@
+| 3    | *Outside trigger,* retrieve secrets when notified from Follower by webhook registered API | *Concern*<br />Follower does not save state and cannot save the webhook URL | - Efficient because only do the job when really needed<br />- Immediate update of K8s Secret on secret change in Conjur | - Need development in Conjur side to support notification mechanism to notify only necessary secrets providers<br />- Follower may not be able to reach secrets provider (need to save URL in Master Database)<br />- Significant development required because it requires a new event mechanism on server |
+| 4    | *Outside trigger,* retrieve secrets when notified by websocket connection | Websocket connection is opened and when there is a change on the server, it will send update | - Efficient because only do the job when really needed<br />- Immediate update of K8s Secret on secret change in Conjur | - Need development in Conjur side to support notification mechanism to notify only necessary secrets providers<br />- Very expensive to hold a socket open just for few rare notifications<br />- Significant development required because it requires a new event mechanism on server |
+
+Decision: **Solution #2, Outside trigger using pub/sub message
+**
+
+The Pub/Sub pattern is a pattern where senders publish messages without knowing who is subscribing to those messages. Once an event is emitted or a message is published, all receivers that subscribe to that event will get a notification. For our purposes, we propose to use a 3rd party pub/sub message handlers, ZMQ, the publisher, that will trigger an event whenever a Conjur secret has been updated. The Secrets Provider, the subscriber, will listen on a port for these events/messages. Once a message is received this will prompt the Secrets Provider to update the K8s Secrets with the updated Conjur Secrets.
+
+We chose to use the pub/sub pattern over waiting for a configurable amount of time because this delivers a nicer customer experience. The customer's K8s Secrets will be updated immediately which in turn avoids situations where customer applications fail to connect to their endpoints. 
+
+#### Flow
+
+1. Configure K8s Secrets with ***conjur-map\*** metadata defined 
+
+   - *Optional:* add label to K8s Secret 
+
+2. Create Service Account for Secrets Provider 
+
+3. Create Role / ClusterRole for the Service Accounts with **get**/**update** privileges
+
+4. Create RoleBinding / ClusterRoleBinding and bind the Service Accounts to the role from previous step
+
+5. Create/Deploy Secrets Providers as Pods in deployment manifest(s) **(\*Milestone 2)\***
+
+6. - If label filtering is defined, add "K8S_SECRET_LABEL" in the Secrets Provider manifest(s)
+
+7. When the Secrets Provider receives a message that Conjur secrets have changed, the process of fetching and updating the K8s Secret(s) repeats
```

Comment:
how does it know which k8s secrets it should update? it this the flow:
1. Secret `some-secret` is updated in Conjur
2. secrets-provider is notified
3. secrets-provider iterates on all the k8s secrets that have labels (or are present in the `K8S_SECRETS` env var) 
  3.1 for each such a secret, get the `conjur-map` entry and iterate over it
    3.1.1 for each entry in the conjur-map check if the value is the Conjur secret that was updated.
    3.1.2 if so, update the value of the k8s-secret with the new value

This seems like a huge flow to do for every secret. If a customer has 1000 secrets in Conjur and only 100 of them are in k8s then the secrets-provider will perform a ton of work for nothing.

Reply:
I took out Milestone 2 for now for the sake of simplicity. Let's readdress this when we pursue Milestone 2